producer/buffer/buffer.go

```diff
@@ -97,16 +100,19 @@ func (b *buffer) Add(d producer.Data) (producer.RefCountedData, error) {
 		b.Unlock()
 		return nil, errBufferClosed
 	}
-	var (
-		dataSize      = uint64(d.Size())
-		maxBufferSize = b.maxBufferSize
-	)
-	if dataSize > maxBufferSize {
+	s := d.Size()
+	if s > b.maxMessageSize {
+		b.Unlock()
+		b.m.dataTooLarge.Inc(1)
+		return nil, errDataTooLarge
+	}
+	dataSize := uint64(s)
+	if dataSize > b.maxBufferSize {
```

Comment:
Is there a need to have both?

Reply:
yeah, this check here is to avoid uint calculation overflow

Reply:
I see.